src/main/java/com/redhat/devtools/lsp4ij/features/highlight/LSPTargetElementEvaluator.java

```diff
@@ -0,0 +1,33 @@
+package com.redhat.devtools.lsp4ij.features.highlight;
+
+import com.intellij.codeInsight.TargetElementEvaluatorEx2;
+import com.intellij.codeInsight.highlighting.HighlightUsagesHandlerBase;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LanguageServersRegistry;
+import com.redhat.devtools.lsp4ij.client.ProjectIndexingManager;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+
+public class LSPTargetElementEvaluator extends TargetElementEvaluatorEx2 {
+    @Override
+    public @Nullable PsiElement adjustReferenceOrReferencedElement(@NotNull PsiFile file, @NotNull Editor editor, int offset, int flags, @Nullable PsiElement refElement) {
+        if (!LanguageServersRegistry.getInstance().isFileSupported(file)) {
+            return null;
+        }
+
+        if (ProjectIndexingManager.isIndexingAll()) {
+            return null;
+        }
+
+        // Try to find the highlighted identifier at the caret and return that fake PSI element
```

Comment:
Let me know if you think there's a better way to discern the identifier at the current offset.

Reply:
The main problem with your strategy is that if your language server doesn't support `textDocument/documentHighlight` it will not work.

If I understand correctly your code, you need to create a fake PsiElement according a given offset with a text range. Right?

If it that I do that to support find usages at https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/usages/LSPUsageTargetProvider.java#L69-L82

If you think it is the same case, I suggest that you create a static method in LSP4IJ that you can use in both cases (in find usage target and in your case).